lang/LangSource/SC_CLIOptions.cpp

```diff
@@ -0,0 +1,277 @@
+
+    auto generic_description = build_generic_description();
+    auto terminal_description = build_terminal_description(terminal_options);
+    auto language_config_description = build_language_config_description();
+
+    // Hidden options, will be allowed both on command line and
+    // in config file, but will not be shown to the user.
+    po::options_description hidden_description("Hidden options");
+    hidden_description.add_options()("input-file", po::value<string>(&mInputFile),
+                                     "input file")("sc-args", po::value<vector<string>>(&mScArgs), "sc args");
+
+    // positional options are declared here but captured via hidden description
+    po::positional_options_description pos_options;
+    pos_options.add("input-file", 1);
+    pos_options.add("sc-args", -1);
+
+    po::options_description cmdline_options;
+    cmdline_options.add(generic_description)
+        .add(terminal_description)
+        .add(hidden_description)
+        .add(language_config_description);
+
+    po::variables_map vm;
+    try {
+        store(po::command_line_parser(argc, argv).options(cmdline_options).positional(pos_options).run(), vm);
+        notify(vm);
+    } catch (po::unknown_option& e) {
+        cerr << e.what() << endl;
+        exit_code = 1;
+        return true;
+    } catch (po::invalid_option_value& e) {
+        cerr << e.what() << endl;
+        exit_code = 1;
+        return true;
+    }
+    if (mParsingFailed) {
+        exit_code = 1;
+        return true;
+    }
+
+    if (vm.count("sc-args")) {
+        terminal_options.mArgs = vm["sc-args"].as<vector<string>>();
+    }
+
+    if (mPrintVersion) {
+        cout << "sclang " << SC_VersionString() << " (" << SC_BuildString() << ")" << endl;
+        return true;
+    }
+
+    // this is handled here and not in the description declaration b/c we need access to the declarations
+    if (mPrintHelp) {
+        // construct a options_description w/o hidden positional options
+        po::options_description visible_description("");
+        visible_description.add(generic_description).add(terminal_description).add(language_config_description);
+
+        cout << "Start a REPL interpreter:           sclang [ -options ]\n"
+                "Execute a .scd file:                sclang [ -options ] path_to_sc_file [ args passed to sclang... ]\n"
+             << visible_description << endl;
+        return true;
+    }
+
+    if (!mInputFile.empty()) {
+        terminal_options.mDaemon = true;
+    }
+
+    SC_LanguageConfig::readLibraryConfig(terminal_options.mStandalone);
+
+    // for the language config we don't use the notifier of program options because
+    // we want to fill the config with a provided yaml file or their default values.
+    // as po does not allow to specify the "notify order", we parse the config manually
+    parse_language_config(vm);
+
+
+    return false;
+}
+
+po::options_description CLIOptions::build_generic_description() {
```

Comment:
`CLIOptions::build_*` methods are really pure functions and as such they should either be `static` methods or free functions! Personally, I tend to prefer the latter.

Reply:
Thanks for the comprehensive review, helps me a lot to learn these things (somehow forgot to pass the strings by reference...).

Concerning this comment tho: these methods need access to the CLIOptions, see e.g.
https://github.com/supercollider/supercollider/blob/6585c12efd2750853f91e709aaf472df743e8ca9/lang/LangSource/SC_CLIOptions.cpp#L97-L102
where `mPrintVersion` and `mPrintHelp` are member variables.

Where it was possible to keep them pure I used static, but I failed to come up with a way to make them static here. So, is it fine to use non-static methods here?

Reply:
> where mPrintVersion and mPrintHelp are member variables.

A sorry, totally missed that! Nevermind then.